### PR TITLE
dialects: Reduction support for scf parallel

### PR DESCRIPTION
**Pull request by mesham**
Adds the reduce and reduce return operations to the scf dialect and supports initVals (and results) from scf parallel. Enhanced verify methods based of scf parallel based upon MLIR constraints

### CODE REVIEW

File: tests/filecheck/mlir-conversion/with-mlir/dialects/scf/parallel_with_reduce.mlir [hunk 1]
```diff
@@ -0,0 +1,36 @@
+// RUN: xdsl-opt %s -t mlir | mlir-opt --mlir-print-op-generic | xdsl-opt -f mlir -t mlir | filecheck %s
```
Comment (PapyChacal) [inline tests/filecheck/mlir-conversion/with-mlir/dialects/scf/parallel_with_reduce.mlir:1]: Di you try without `-t mlir`? I think it is the default now.

**Review by PapyChacal — APPROVED** [review]
I have some nitpicking suggestions but no strong feeling about them; happy to merge as-is or with those adressed!

File: xdsl/dialects/scf.py [hunk 2]
```diff
@@ -147,29 +147,187 @@ def get(
...
                 f"{len(self.step)}.")
-        body_args = self.body.block.args if len(self.body.blocks) != 0 else ()
-        if len(self.lowerBound) != len(body_args) or not all(
-            [isinstance(a.typ, IndexType) for a in body_args]):
+
+        # An scf.parallel operation must have one block only
+        if len(self.body.blocks) != 1:
+            raise VerifyException(
+                "scf.parallel region must contain exactly 1 block but "
+                f"{len(self.body.blocks)} were provided")
```
[the remark this answers is not in the document]
Comment (mesham) [reply xdsl/dialects/scf.py]: Thanks, good spot - I've added this in and removed that validation check as it is now checked elsewhere

File: xdsl/dialects/scf.py [hunk 3]
```diff
@@ -147,29 +147,187 @@ def get(
...
+                    f", parallel argment is of type {initValsType} whereas reduction operation is of type {typ}"
+                )
+
+        # Ensure that scf.yield is the last operation in the block as this is required
+        if len(self.body.block.ops) == 0 or not isinstance(
+                self.body.block.ops[-1], Yield):
+            raise VerifyException(
+                "scf.parallel region must terminate with an scf.yield")
+
+        # Ensure that the number of operations in scf yield is exactly zero
```
[the remark this answers is not in the document]
Comment (mesham) [reply xdsl/dialects/scf.py]: Fixed :)

File: xdsl/dialects/scf.py [hunk 4]
```diff
@@ -147,29 +147,187 @@ def get(
...
+        number_yield_ops = len(self.body.block.ops[-1].arguments)
+        if number_yield_ops != 0:
+            raise VerifyException(
+                f"scf.yield contains {number_yield_ops} operands but this must be zero"
+            )
```
[the remark this answers is not in the document]
Comment (mesham) [reply xdsl/dialects/scf.py]: Fixed :)

File: xdsl/dialects/scf.py [hunk 5]
```diff
@@ -147,29 +147,187 @@ def get(
...
+        block: Block,
+    ) -> ReduceOp:
+        return ReduceOp.build(operands=[argument],
+                              regions=[Region(block, parent=None)])
+
+    def verify_(self) -> None:
+        if len(self.body.blocks) != 1:
+            raise VerifyException(
+                "scf.reduce region must contain exactly 1 block but "
+                f"{len(self.body.blocks)} were provided")
```
[the remark this answers is not in the document]
Comment (mesham) [reply xdsl/dialects/scf.py]: Thanks, I've moved to singleBlockRegion and removed that validation check. As we discussed, this does look like a restriction based on the opening paragraph of the MLIR reduction operation description (which says it must contain one block with two arguments) - but then talking about the entry block later on it seems to slightly contradict itself.

File: tests/dialects/test_scf.py [hunk 6]
```diff
@@ -56,6 +57,352 @@ def test_parallel():
...
+def test_parallel_test_count_number_reduction_ops():
+    reductions: List[ReduceOp] = []
+    for i in range(10):
+        init_val = Constant.from_int_and_width(i, i32)
+        reduce_op = ReduceOp.get(init_val, Block())
+        reductions.append(reduce_op)
+
+    b = Block()
+    b.add_ops(reductions)
+    p = ParallelOp.get([], [], [], Region(b))
```
Comment (superlopuh) [inline tests/dialects/test_scf.py:278]:
```suggestion
    @Builder.implicit_region
    def body():
        for i in range(10):
            init_val = Constant.from_int_and_width(i, i32)
            ReduceOp.get(init_val, Block())
p = ParallelOp.get([], [], [], body)
```
This should create exactly the same IR

File: tests/dialects/test_scf.py [hunk 7]
```diff
@@ -1,9 +1,10 @@
-from typing import cast
-
+import pytest
+from typing import cast, List
```
Comment (superlopuh) [inline tests/dialects/test_scf.py:2]: nit: prefer `list` over `List`

File: tests/dialects/test_scf.py (hunk 6 above)
Comment (superlopuh) [reply tests/dialects/test_scf.py:278]: Similarly below

Comment (mesham) [reply tests/dialects/test_scf.py]: Thanks, that's great - fixed :)

File: tests/dialects/test_scf.py (hunk 7 above)
Comment (mesham) [reply tests/dialects/test_scf.py]: Fixed :) I don't need List now that we are using the builder so I have removed it from the import :)

**Review by superlopuh — APPROVED** [review]

**Pull request merged**